Support HTTP streams

Shiranuit opened the pull request:
⚠️ #2266 Should be merge in this PR before merging this PR into 2-dev

## What does this PR do ?

This PR add the capability to send stream of data through the HTTP protocol,
this way we can later implement download and data processing on the fly without storing everything in RAM.

From now on returning an instance of HttpStream as a result of a controller action will make Kuzzle send data
through HTTP chunk by chunk.

Kuzzle supports streaming with a fixed and dynamic size.
Fixed size can be achieved by setting the `totalBytes` property of the `HttpStream` instance

### Examples

```js
const fs = require('fs');
async myAction (request) {
 const readStream = fs.createReadStream('./Photos.zip');
 return new HttpStream(readStream);
}
```

```js
const { PassThrough } = require('stream');
async myAction (request) {
 const stream = new PassThrough();
 setInterval(() => {
   // Process data
   stream.write('data');
  if (done) {
    stream.end();
  }
 }, 50):
 return new HttpStream(stream);
}
```
### How should this be manually tested?

`npm run test` 

For now it's pretty hard to test this feature functionally without having any controller action using it.

### To Do
- [x] Unit Tests
- [x] Documentation

Aschen commented on doc/2/guides/advanced/streams/index.md:
```diff
@@ -0,0 +1,57 @@
+order: 300
+---
+
+# HTTP Streams <SinceBadge version="auto-version" />
```
This section should rather be in the **Develop On Kuzzle** > **API Controller** section (We have a [Response Format](https://docs.kuzzle.io/core/2/guides/develop-on-kuzzle/api-controllers/#response-format) section)

Shiranuit replied:
I think this section might be useful here too if someone stumble uppon this page, what do you think ?
should we duplicate this section or juste move it ?

Aschen replied:
IMHO it's not a very common usage to manipulate HTTP streams.  Streams are more a subset of what you can do when defining API action than an entire feature on it's own.  
About duplication, it's generally not a good idea I think

Aschen commented on doc/2/guides/advanced/streams/index.md:
````diff
@@ -0,0 +1,57 @@
+
+ return new HttpStream(stream);
+}
+```
````
IMHO this example is confusing, maybe you could replace it with a stream of CSV lines from SearchResult pagination request?

Shiranuit replied:
I agree, I will replace it with a better example once I get the time

Aschen commented on lib/api/request/kuzzleRequest.ts:
```diff
@@ -246,6 +246,10 @@ export class KuzzleRequest {
     }
 
+    if (this.context.connection.protocol !== 'http' && result instanceof HttpStream) {
+      throw new InternalError('cannot set an HttpStream as a request\'s response in a non-http protocol');
```
You should define a new error in the error manager rather than throwing a raw error

Aschen commented on lib/core/network/protocols/httpwsProtocol.js:
```diff
@@ -636,6 +623,160 @@ class HttpWsProtocol extends Protocol {
+  httpSendStream (request, response, httpStream, message) {
+    const streamSizeFixed = typeof httpStream.totalBytes === 'number' && httpStream.totalBytes > 0;
+
+    if (httpStream.stream._readableState.errored
```
Nit
```suggestion
    if ( httpStream.stream._readableState.errored
```

Aschen commented on lib/core/network/protocols/httpwsProtocol.js:
```diff
@@ -636,6 +623,160 @@ class HttpWsProtocol extends Protocol {
+    });
+    
+
+    httpStream.stream.on('error', err => {
```
What happen to the response if the user stream is destroyed? Shouldn't we `end` it with an error?

Shiranuit replied:
Yes you are right I forgot to end the stream in case of stream failure

Aschen reviewed (changes requested):
:tada:

Shiranuit commented on doc/2/framework/classes/http-stream/properties/index.md:
```diff
@@ -0,0 +1,25 @@
+
+| Type                                                                         | Description                |
+|------------------------------------------------------------------------------|----------------------------|
+| <pre>[BackendCluster](/core/2/framework/classes/backend-cluster)</pre> | BackendCluster instance |
```
Oops 😅 

And yes stream and totalBytes are readonly properties that can only be set at construction

Shiranuit commented on lib/core/network/protocols/httpwsProtocol.js:
```diff
@@ -657,13 +657,18 @@ class HttpWsProtocol extends Protocol {
     const streamSizeFixed = typeof httpStream.totalBytes === 'number' && httpStream.totalBytes > 0;
 
     if (httpStream.stream._readableState.errored || httpStream.stream._readableState.closed) {
+      response.writeStatus('200 OK');
```
That's right I totally forgot to change it  😅

Aschen commented on doc/2/guides/advanced/streams/index.md:
```diff
@@ -0,0 +1,57 @@
+const fs = require('fs');
+
+async myAction (request) {
+ const readStream = fs.createReadStream('./Document.zip');
```
Too much Windows for you @Shiranuit

Aschen reviewed (approved):
Read my comment about the documentation, apart from that it's really fine work :+1: 

(Also I think an example with sending documents as CSV from search result could be a better example)

Shiranuit commented:
@Aschen 
> (Also I think an example with sending documents as CSV from search result could be a better example)

After implementing it in document:export I'm not sure that this is a good example, it's not that easy to do it properly and I do not want to provide a bad example that could lead to bad usage of this.

Shiranuit commented:
As request I moved the documentation and I added the following warning
```
::: warning
Chunks are send through the HTTP Protocol each time a chunk is emitted through the `data` event of the given stream.
It's up to you to implement a buffer mechanism to avoid sending too many small consecutive chunks through the network.

Sending too many small chunks instead of bigger chunks will increase the number of syscall made to the TCP Socket and might decrease performance and throughput.
:::
```

Aschen commented on doc/2/guides/develop-on-kuzzle/api-controllers/index.md:
```diff
@@ -585,6 +585,45 @@ app.controller.register('redirect', {
+Sending too many small chunks instead of bigger chunks will increase the number of syscall made to the TCP Socket and might decrease performance and throughput.
+:::
+
+## Usage
```
Wrong heading section.

Can you use the same format as other examples? 
![image](https://user-images.githubusercontent.com/4447392/153161873-9655a70f-7332-4a98-85fe-f777a1bf42d7.png)

Shiranuit replied:
@Aschen Yes I fixed the documentation in #2266

Shiranuit replied:
I copied back the changes made to the documentation from #2266
